### PR TITLE
[MM-12396] Option to add user to a channel from the profile pop-over (permissions fix)

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This fixes a permission issue for the Add User to Channel modal button in the profile popover. The button was not showing up for Team Admins or Channel Admins when toggled in the System Console. 

Since toggling `manage_x_channel_members` permissions affects users of different roles through their *channel* permissions and not *team* permissions, we can check if the user has `manage_x_channel_members` permission in any channel within the current team using `haveIChannelPermission` on the available channel memberships. If they have permissions in at least one channel, the Add User to Channel appears in the Profile Popover.

[Relevant PR](https://github.com/mattermost/mattermost-webapp/pull/1918) for this fix

#### Ticket Link
[Github Ticket](https://github.com/mattermost/mattermost-server/issues/9501)
[Jira Ticket](https://mattermost.atlassian.net/browse/MM-12396)

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has redux changes (please link)

[Redux PR](https://github.com/mattermost/mattermost-redux/pull/806)

### CODE REVIEW

**Comment by mickmister** [comment]
@jwilander here's a PR we can use to deploy with the test license

**Comment by jwilander** [comment]
@mickmister license uploaded

File: components/profile_popover/index.js [hunk 1]
```diff
@@ -21,6 +29,34 @@ import {getSelectedPost, getRhsState} from 'selectors/rhs';
 
 import ProfilePopover from './profile_popover.jsx';
 
+function hasManageChannelMemberPermission(state) {
```
Comment (mickmister) [inline components/profile_popover/index.js]: I'm thinking maybe this should be in a `selectors` file instead?

**Comment by mickmister** [comment]
@hanzei this is ready for review now. test server will need to be re-applied again

**Comment by mickmister** [comment]
@lindy65 The permissions issue should be resolved by this PR. The new spinmint server needs a test license, however.

**Comment by lindy65** [comment]
Thanks @mickmister,

I've added a test licence to the spinmint

Checking I understand correctly that the "Add to a channel" feature is using `haveIChannelPermission` to check permissions and that the user adding another user to the channel needs permission to do so in that channel? (as per your intro here):

> haveIChannelPermission is used in the mapStateToProps instead of using a TeamPermissionGate.

1. Confirm that the option to "Add to a channel" does not appear in another user's profile pop-over for a normal user who does not have permission to `manage channel members` 

2. Confirm that while the permission in 1) is still in place and I promote the test user to a "Team Admin" and change `team admin` permissions to allow `manage channel members`, the newly promoted Team Admin user sees the "Add to a channel" option in another user's profile pop-over.

3. Confirm that while the permission in 1) is still in place and I demote the test user to a "Channel Admin" of a specific channel and change `channel admin` permissions to dis-allow `manage channel members`, the newly demoted Channel Admin does NOT see the "Add to a channel" option in another user's profile pop-over from within the specific channel.

4. I am seeing an issue where the profile pop-over of a user remains in front of the channel selection modal which makes it difficult to verify the channel name.  The profile pop-over disappears only at the time of clicking on the channel name in the filtered results.  However, this also occurs on `master`.  Perhaps the profile pop-over can be made to close once you start typing in the text input box for "Add [member_name] to a channel"? Unsure whether this is out of scope for this PR?

image

5. I'm also seeing an issue in the current user's channel sidebar where, if they have a channel as an unread channel in their channel sidebar and they add another user to that channel from the profile pop-over, the unread channel in their view becomes read and is removed from "Unreads" section - see .gif  This also occurs on `master` so I am not sure whether it is out of scope for this PR?

unread channel in current user's sidebar disappears when adding another user to that channel

Let me know if you have any questions 👍 (All the permissions stuff gets rather confusing and is difficult to explain in a written message so feel free to let me know if anything is unclear).

**Comment by mickmister** [comment]
1.

> Confirm that the option to "Add to a channel" does not appear in another user's profile pop-over for a normal user who does not have permission to manage channel members

Confirmed

2.

> Confirm that while the permission in 1) is still in place and I promote the test user to a "Team Admin" and change team admin permissions to allow manage channel members, the newly promoted Team Admin user sees the "Add to a channel" option in another user's profile pop-over.

Confirmed

3.

> Confirm that while the permission in 1) is still in place and I demote the test user to a "Channel Admin" of a specific channel and change channel admin permissions to dis-allow manage channel members, the newly demoted Channel Admin does NOT see the "Add to a channel" option in another user's profile pop-over from within the specific channel.

Confirmed

Just to be clear, if the Channel Admin user is given the Manage Channel Member permissions in the system console, they will see the "Add to a channel" button if they have the Channel Admin status in *any* channel of the team (as long as the private/public flag agrees with the set permission). The current channel they are viewing is also irrelevant to whether they can see the button in the profile popover. Also, they will only see channels in the drop down of the modal that they have permissions to manage.

4.

> I am seeing an issue where the profile pop-over of a user remains in front of the channel selection modal which makes it difficult to verify the channel name. The profile pop-over disappears only at the time of clicking on the channel name in the filtered results. However, this also occurs on master. Perhaps the profile pop-over can be made to close once you start typing in the text input box for "Add [member_name] to a channel"? Unsure whether this is out of scope for this PR?

I've found the issue with this. When opening the profile popover with an "@ mention", versus clicking on their profile picture on the left side of a post, the "@ mention" scenario does not supply a function to the popover to allow itself to close. You can see this happen if you "@ mention" yourself, and open the "Edit Account Settings" modal. You should see the popover stay open. Likewise, if the popover for another user is opened by clicking on their profile picture, the "Add to a Channel" modal does correctly close the popover on open. The fix for this is simple, and I'm opening a PR for the fix.

5.

> I'm also seeing an issue in the current user's channel sidebar where, if they have a channel as an unread channel in their channel sidebar and they add another user to that channel from the profile pop-over, the unread channel in their view becomes read and is removed from "Unreads" section - see .gif This also occurs on master so I am not sure whether it is out of scope for this PR?

I've found the issue with the one. When a new member is added to a channel, the backend also marks that channel as read to the current user. See app/channel.go [line 886.](https://github.com/mattermost/mattermost-server/blob/b4d645f121ef71f56a0118de1b4ef376834ea481/app/channel.go#L886) This functionality has unexpected behavior while adding a user from a different channel than the target channel. Since it marks the target channel as viewed, the channel is no longer bold nor showing a badge. I'm not entirely sure what the purpose of this line is. This will need some discussion around next steps, and I'd say it's out of scope for this PR.

**Comment by lindy65** [comment]
Hi @mickmister :)

Thank you for your detailed feedback!

All good on 1), 2) and 3) and thank you for the additional explanatory note on 3)

> Just to be clear, if the Channel Admin user is given the Manage Channel Member permissions in the system console, they will see the "Add to a channel" button if they have the Channel Admin status in any channel of the team (as long as the private/public flag agrees with the set permission). The current channel they are viewing is also irrelevant to whether they can see the button in the profile popover. Also, they will only see channels in the drop down of the modal that they have permissions to manage.

4) Thank you for finding the cause of the issue and for working on a PR to fix it!

5) For this one, perhaps next step is for me to create a Jira ticket for investigation but I'll defer to @jwilander or @saturninoabril to guide us here.

**Comment by saturninoabril** [comment]
Thanks @mickmister @lindy65, I agree on item 5 to be a separate issue.  Please file a Jira ticket to investigate/fix and directly assign to me.

**Comment by lindy65** [comment]
Thanks @saturninoabril - I have filed https://mattermost.atlassian.net/browse/MM-14588 to address the issue and have assigned it to you.

**Comment by lindy65** [comment]
Thanks @saturninoabril 

[Here](https://mattermost.atlassian.net/browse/MM-14588) is the ticket for the issue in 5)

**Comment by esethna** [comment]
@lindy65 @saturninoabril @mickmister is this PR ready for review?

**Comment by mickmister** [comment]
@esethna ready for review 👍

**Comment by esethna** [comment]
Thanks @lindy65 has this passed QA review?

**Comment by lindy65** [comment]
Yes, @esethna this PR has passed QA review... 👍🏻

**Comment by mickmister** [comment]
Copying @jespino as we had some discussion on the implementation here

File: components/profile_popover/index.js (hunk 1 above)
Comment (jespino) [reply components/profile_popover/index.js]: Yes, I think so

File: components/profile_popover/index.js [hunk 2]
```diff
@@ -21,6 +29,34 @@ import {getSelectedPost, getRhsState} from 'selectors/rhs';
...
+    const myMemberships = getMyChannelMemberships(state);
+    const teamId = getCurrentTeamId(state);
+    const permissions = [
+        Permissions.MANAGE_PUBLIC_CHANNEL_MEMBERS,
+        Permissions.MANAGE_PRIVATE_CHANNEL_MEMBERS,
+    ];
+
+    for (const channelId in myMemberships) {
+        if (myMemberships.hasOwnProperty(channelId)) {
+            for (const permission of permissions) {
```
Comment (jespino) [inline components/profile_popover/index.js]: I think the for loop here for the two permissions is not needed I would use something like:
```js
if (haveIChannelPermission(state, {channel: channelId, team: teamId, Permissions.MANAGE_PUBLIC_CHANNEL_MEMBERS}) ||
    haveIChannelPermission(state, {channel: channelId, team: teamId, Permissions.MANAGE_PRIVATE_CHANNEL_MEMBERS}) {
    return true;
}
```

File: components/profile_popover/index.js [hunk 3]
```diff
@@ -21,6 +29,34 @@ import {getSelectedPost, getRhsState} from 'selectors/rhs';
...
 
+function hasManageChannelMemberPermission(state) {
+    const myMemberships = getMyChannelMemberships(state);
+    const teamId = getCurrentTeamId(state);
+    const permissions = [
+        Permissions.MANAGE_PUBLIC_CHANNEL_MEMBERS,
+        Permissions.MANAGE_PRIVATE_CHANNEL_MEMBERS,
+    ];
+
+    for (const channelId in myMemberships) {
```
Comment (jespino) [inline components/profile_popover/index.js]: A more explicit wat that no require the "hasOwnProperty" check would be something like:
```js
for (const channelId of Object.keys(myMemberships)) {
    ...
}
```

**Comment by mickmister** [comment]
@jespino suggested changes have been added.

As for the choosing where the `canManageAnyChannelMembers` selector should go, here is what it looks like if it were to exist in the mattermost-redux repo [here](https://github.com/mickmister/mattermost-redux/commit/a76f5680d2fac87efc5c81e4ef3cdb0532f3202a). I think it would make sense to have it in the mattermost-redux repo, as it seems appropriate to have unit tests for it there. I'm wondering if other files will want to reuse this functionality in the future. What are your thoughts?

**Comment by jespino** [comment]
@mickmister yes, mattermost-redux should be the place for that, is generic enough to affect to other clientes.

**Comment by mickmister** [comment]
This PR is complete. The PR for Redux is [here](https://github.com/mattermost/mattermost-redux/pull/806).

**Comment by esethna** [comment]
Is this PR dependent on the redux one? I'm getting a JS error when trying to open the user pop-over.

image

**Comment by mickmister** [comment]
@esethna Yes it is dependent on the redux PR. I think the redux PR would need to be merged before this is testable here. Is there usually a solution here when this is the case?

**Comment by mickmister** [comment]
Actually, that error is happening for a different reason. The memberships being used to determine permissions are for every team on the server. There is a channel lookup on each membership to determine the channel type, and the available channels are scoped to the current team, as it should be.

There needs to be a check for only looping through the current team's memberships. I'll need to circle back to this and make this correction. Thanks for catching this @esethna

**Comment by mickmister** [comment]
Redux PR is merged. This is ready for review @esethna

**Comment by esethna** [comment]
@lindy65 can we do another QA pass on this PR since there were changes made?

**Comment by lindy65** [comment]
Thanks @mickmister @esethna 

I added an E20 licence to the spinmint and tested:
- Previous issues 1-4 above and did not see any problems
- the areas fixed by https://github.com/mattermost/mattermost-redux/pull/806 as follows:
  - Test1: If manage channel member permissions is disabled for all members but manage public channels is enabled for Channel Admins
  - Result1: the Channel Admin has permission to add members to public channels on the current team
  - Test2: If manage channel member permissions is disabled for all members but manage private channels is enabled for Team Admins
  - Result2: the Team Admin will have permission to add members to private and public channels on the current team

Please confirm test results are indeed what is expected and let me know whether there are any other scenarios that need testing.

**Comment by mickmister** [comment]
@lindy65 @esethna 

All test results are passing for me as well on the spinmint server. This is ready for dev review.

The issue 5 listed above pertaining to clearing unreads in other channels still exists. I'm not sure if we want to proceed with merging before this is resolved. What are your thoughts? @lindy65 @saturninoabril

**Comment by saturninoabril** [comment]
Thanks @mickmister, re: item 5, we can address it separately.  Associated ticket is scheduled for 5.11.

**Comment by mickmister** [comment]
@jespino This PR should be good to go, can you give this another review when you have the chance?

Also, I'm not sure what to do about the circleci build error. The printed error says it's related to project config.

**Comment by hanzei** [comment]
@mickmister You can ignore the circleci error. We are currently in the process of setting this up.

**Comment by jespino** [comment]
Thanks @mickmister!!

**Pull request merged**